src/GNDStk/gnds-tree-node.hpp

```diff
@@ -0,0 +1,444 @@
+         error(
+            "Node child() called with key \"" + str + "\", "
+            "but this key wasn't\nfound in the node's children."
+         );
+      }
+
+      // done
+      return empty;
+   }
+
+   // for child_t<T>
+   template<class T, class META, class CHILD>
+   auto child(const child_t<T,META,CHILD> &kwd) const
+   {
+      return tnode<MCON,CCON,T>(child(kwd.name));
+   }
+
+   // for child_t<variant>, caller must stipulate <T>.
+   template<class T, class META, class CHILD, class... Ts>
+   auto child(const child_t<std::variant<Ts...>,META,CHILD> &kwd) const
+   {
+      return tnode<MCON,CCON,T>(child(kwd.name));
+   }
+
+
+   // ------------------------
+   // operator()
+   // ------------------------
+
+   // These provide notationally consistent access to the above meta() and
+   // child() functions, except that the std::string parameter versions of
+   // those functions are intentionally not supported here, as it would be
+   // indeterminate as to whether we'd want the meta or child (std::string)
+   // function. This is not a deficiency, but instead reflects the fact that
+   // calling these functions with meta_t and child_t parameters - not with
+   // std::string parameters - should be preferred. Those encode (via their
+   // type) whether we're accessing metadata or children.
+
+   // meta_t
+   // forwards to meta(meta_t) above
+   template<class T>
+   decltype(auto) operator()(const meta_t<T> &kwd) const
+   {
+      return meta(kwd);
+   }
+
+   // child_t
+   // forwards to child(child_t) above
+   template<class T, class META, class CHILD>
+   decltype(auto) operator()(const child_t<T,META,CHILD> &kwd) const
+   {
+      return child(kwd);
+   }
+
+   // child_t, ...
+   // multi-argument
+   template<class T, class META, class CHILD, class... Ts>
+   decltype(auto) operator()(
+      const child_t<T,META,CHILD> &kwd,
+      Ts &&...ts
+   ) const {
+      return (*this)(kwd)(std::forward<Ts>(ts)...);
+   }
+
+}; // class Node
+
+
+
+// Node::static_found
+template<
+   template<class...> class MCON,
+   template<class...> class CCON
+>
+bool Node<MCON,CCON>::static_found = false;
+
+
+
+// -----------------------------------------------------------------------------
+// write
+// -----------------------------------------------------------------------------
+
+// write(char *)
+template<
+   template<class...> class MCON,
+   template<class...> class CCON
+>
+inline bool Node<MCON,CCON>::write(
+   const char * const file,
+   const int level
+) const {
+   // calls write(ostream) below
+   std::ofstream ofs(file);
+   return !write(ofs,level).fail();
+}
+
+
+// write(ostream)
+template<
+   template<class...> class MCON,
+   template<class...> class CCON
+>
+std::ostream &Node<MCON,CCON>::write(
+   std::ostream &os,
+   const int level
+) const {
+   // indentation
+   const std::string icurr(indent* level   ,' ');
+   const std::string inext(indent*(level+1),' ');
```

Comment:
This is a pattern I've used occasionally in other codes. It could be done in any number of ways, but I think this way is more clear (or less unclear) than other ways of doing it. Node's write function indents its output depending on where the node is in the overall tree. "level" is the indentation level: 0 for no indentation, 1 for indent one level, 2 for indent two levels, etc. "indent" is the number of spaces you want in each indentation level. Many people like 2-space indentation; I like 3-space. (My code allows users to set this.) Multiply the indentation level by the number of spaces per level, and you get, well, the total number of spaces. Turns out that in the function in question, I use the current level (level) when indenting the node's name, and the next level (level+1) when indenting the node's children. Recurse with level+1. One could achieve the same effect in any number of ways; this is a nice local way that doesn't involve passing strings around or modifying them. I wouldn't worry about the apparent less efficiency it might have; output to disk (or even an internal location) would be the bottleneck here. I'll add a couple of comments for these two lines.